### ISSUE TITLE
[FR] Fully support PyTorch schemas / signatures in Mlflow Models

### ISSUE THREAD

**Issue by IngoStatworx** [mlflow/mlflow#6820]
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

I propose to modify the PyTorch wrapper here:

https://github.com/mlflow/mlflow/blob/6eccc69286b7d8b34ce6d34a20bf821e8c69feba/mlflow/pytorch/__init__.py#L738

to allow for arbitrary inputs to the model. Sufficient would be to allow for dictionaries, lists or generally iterables of tensor types - in particular iterables of tensors as defined by Mlflow's own model signature logic.

Ideally, if one can define it as a signature in tensor format, the PyTorch wrapper should be able to process it!

It doesnt't seem that the limitation of one array in - one array out holds for other model flavors. As such it appears that this is not a limitation by some other part of the code. It may thus be worth checking whether the change could be as simple as adding a few more input checks or unwrapping operations in the wrapper.

For instance, it does not seem to be necessary to not accept pytorch tensors, when casting to tensors will be done anyway. Furthermore, if the input will be type checked, one could recursively do the same if an iterable type (list or dict) is provided instead of a single input.

### Motivation

> #### What is the use case for this feature?

Mlflow allows considerable flexibility to define a model's signature. Supporting PyTorch is obviously something the Mlflow team strives for (as indicated by the website).

However, the PyTorch wrapper will straight up fail if
1. The input is not a numpy or pandas type - it will even fail if fed a pytorch tensor!
2. An iterable of tensor types
3. Input / Output schemas that would normally be supported by other Mlflow model flavors.

These limitations are not mentioned for other model flavors.

For PyTorch specifically, this is an issue. Increasingly many use cases involve the use of pre-trained models, for instance from HuggingFace.
HuggingFace models specifically take a dict-like input, and thus can not be used.
Similar cases can be made for more complex DL models, for instance

-> Language models with masking tensors
-> Transformer models with memory embeddings of some sort
-> Tabular transformer-models with categorical / continuous tensor inputs (that are embedded differently)
->  Contrastive learning models
-> Retrieval models with paired attention or cross attention
-> Neural search, retrieval enhanced models
-> and many, many more

> #### Why is this use case valuable to support for MLflow users in general?

PyTorch is now the dominant machine learning framework for deep learning models. Mlflow users will want to use deep learning models which typically require more flexible input and output signatures.

> #### Why is this use case valuable to support for your project(s) or organization?

We seek to deploy and train PyTorch models and immediately ran into this issue.

> #### Why is it currently difficult to achieve this use case?

Few SOTA models natively run a single input tensor and a single output tensor. 

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [X] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [X] `integrations/azure`: Azure and Azure ML integrations
- [X] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

**Comment by BenWilson2** [comment]
Hi @IngoStatworx thanks for the idea! We'd be thrilled to review your PR when you submit it. In the meantime, if you have any questions, please let us know how we can help! :)

**Comment by IngoStatworx** [comment]
> Hi @IngoStatworx thanks for the idea! We'd be thrilled to review your PR when you submit it. In the meantime, if you have any questions, please let us know how we can help! :)

Here is a draft PR of something that could work.
I was unable to make the DEV environment work locally, but perhaps you can have a look whether you can use this

https://github.com/mlflow/mlflow/pull/6851
